provider/cluster/hostname.go

```diff
@@ -0,0 +1,226 @@
+			blockedHostnames = append(blockedHostnames, name[1:])
+		} else {
+			blockedHostnames = append(blockedHostnames, name)
+		}
+	}
+
+	hs := &hostnameService{
+		inUse:            make(map[string]dtypes.DeploymentID),
+		blockedHostnames: blockedHostnames,
+		blockedDomains:   blockedDomains,
+		requests:         make(chan reserveRequest),
+		releases:         make(chan []string),
+		lc:               lifecycle.New(),
+	}
+
+	go hs.lc.WatchContext(ctx)
+	go hs.run()
+
+	return hs
+}
+
+func (hs *hostnameService) run() {
+	defer hs.lc.ShutdownCompleted()
+
+loop:
+	for {
+
+		// Wait for any service to finish
+		select {
+		case <-hs.lc.ShutdownRequest():
+			hs.lc.ShutdownInitiated(nil)
+			break loop
+		case rr := <-hs.requests:
+			hs.doRequest(rr)
+		case hostnames := <-hs.releases:
+			hs.doRelease(hostnames)
+		}
+	}
+
+}
+
+var errHostnameNotAllowed = errors.New("hostname not allowed")
+
+func (hs *hostnameService) isHostnameBlocked(hostname string) error {
+	for _, blockedHostname := range hs.blockedHostnames {
+		if blockedHostname == hostname {
+			return fmt.Errorf("%w: %q is blocked by this provider", errHostnameNotAllowed, hostname)
+		}
+	}
+
+	for _, blockedDomain := range hs.blockedDomains {
+		if strings.HasSuffix(hostname, blockedDomain) {
+			return fmt.Errorf("%w: domain %q is blocked by this provider", errHostnameNotAllowed, hostname)
+		}
+	}
+
+	return nil
+}
+
+func (hs *hostnameService) doRequest(rr reserveRequest) {
+	// check if hostname is blocked
+	for _, hostname := range rr.hostnames {
+		hostname = strings.ToLower(hostname)
+		blockedErr := hs.isHostnameBlocked(hostname)
+		if blockedErr != nil {
+			rr.result <- blockedErr
+			return
+		}
+		takenBy, hostnameTaken := hs.inUse[hostname]
+		if hostnameTaken && !takenBy.Equals(rr.dID) {
+			rr.result <- fmt.Errorf("%w: %q already in use", errHostnameNotAllowed, hostname)
+			return
+		}
+	}
+
+	if rr.doReserve {
+		for _, hostname := range rr.hostnames {
+			hs.inUse[hostname] = rr.dID
+		}
+	}
+
+	rr.result <- nil // No error
+}
+
+func (hs *hostnameService) doRelease(hostnames []string) {
+	for _, hostname := range hostnames {
+		delete(hs.inUse, hostname)
+	}
+}
+
+func (hs *hostnameService) ReserveHostnames(hostnames []string, did dtypes.DeploymentID) <-chan error {
+	returnValue := make(chan error, 1) // Buffer of one so service does not block
+	request := reserveRequest{
+		hostnames: hostnames,
+		result:    returnValue,
+		doReserve: true, // reserve hostnames
+		dID:       did,
+	}
+
+	select {
+	case hs.requests <- request:
+
+	case <-hs.lc.ShuttingDown():
+		returnValue <- ErrNotRunning
+	}
+
+	return returnValue
+}
+
+func (hs *hostnameService) ReleaseHostnames(hostnames []string) {
+	hs.releases <- hostnames
+}
+
+func (hs *hostnameService) CanReserveHostnames(hostnames []string, did dtypes.DeploymentID) <-chan error {
+	returnValue := make(chan error, 1) // Buffer of one so service does not block
+	request := reserveRequest{
+		hostnames: hostnames,
```

Comment:
Maybe `strings.ToLower` it here so the burden is on the calling threads.  Probably doesn't make much of a difference tho.